### PR TITLE
Add `cloneInputAst` option to `babel.transformFromAst`

### LINKED ISSUE #10231: A way to disable cloneDeep in a @babel/core transformFromAstSync
## Feature Request

**Is your feature request related to a problem? Please describe.**
I have an issue with `@babel/core` when I’m using [transformFromAstSync](https://babeljs.io/docs/en/next/babel-core.html#transformfromastsync) [cloneDeep is used](https://github.com/babel/babel/blob/a9a7f4e690b15d569deb8a53719f37d6dc09393b/packages/babel-core/src/transformation/normalize-file.js#L79).
I’m passing `AST` generated by `babel` with help of [recast](https://github.com/benjamn/recast), which stores information about code structure in `prototype` of objects, and this information is lost after transform.

**Describe the solution you'd like**
Would be great to have ability to pass an option  `deepCopy` or something like this, to prevent AST from copying.

**Describe alternatives you've considered**
Alternative solution would be copy `AST` in a compatible with `recast` manner.

**Teachability, Documentation, Adoption, Migration Strategy**

```js
const sourceCode = "if (true) return;";
const parsedAst = babel.parse(sourceCode, { allowReturnOutsideFunction: true });
const { code, map, ast } = babel.transformFromAstSync(parsedAst, sourceCode, {
  ast: true,
  deepCopy: false,
});
```

I’m working on pluggable code transformer [putout](https://github.com/coderaiser/putout), which have it’s own [plugins api](https://github.com/coderaiser/putout#plugins-api), but I have an idea to add support of babel plugins as well, and don’t want to fork `@babel/core` or relay on implementation details for this purpose. The main feature of `putout` is ability to change as small code as possible and keep code style as it is, but after `transformFromAstSync` there is a full file change with a lot modifications here in there.

Idea to support babel plugins came to me after I saw [babel roadmap](https://babeljs.io/docs/en/roadmap#codemods-for-tc39-proposals) which says that babel going to support codemods, and this is amazing :).

### LINKED ISSUE #11206: transformFromAst: Add option to not clone the AST
## Feature Request

- [x] I would like to work on this feature!

**Is your feature request related to a problem? Please describe.**

https://github.com/babel/babel/issues/7987 / https://github.com/babel/babel/pull/8342 clones the AST by default, but cloning can take a considerable amount of time and the caller might decide it is not needed.

**Describe the solution you'd like**
Add an option to `transformFromAst` to not clone the input ast, because I am fine with mutation.

**Teachability, Documentation, Adoption, Migration Strategy**
```js
transformFromAst(ast, sourceCode, {cloneAst: false});
```

### PR DESCRIPTION
**Pull request by coderaiser**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #10231, fixes #11206 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Added new flag `cloneInputAst`
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/2306
| Any Dependency Changes?  |
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/10241"></a>

### CODE REVIEW

File: packages/babel-core/src/transformation/normalize-file.js [hunk 1]
```diff
@@ -76,7 +76,7 @@ export default function normalizeFile(
     } else if (ast.type !== "File") {
       throw new Error("AST root must be a Program or File node");
     }
-    ast = cloneDeep(ast);
+    if (options.deepCopy) ast = cloneDeep(ast);
   } else {
```
Comment (nicolo-ribaudo) [inline packages/babel-core/src/transformation/normalize-file.js]: Can you add an error here, in case the new option is used with a string input?

Comment (coderaiser) [reply packages/babel-core/src/transformation/normalize-file.js]: Done

**Comment by coderaiser** [comment]
That’s a good idea, `cloneInputAst` feats more 🙂.

**Comment by nicolo-ribaudo** [comment]
@coderaiser I have rebased this PR, since it looks like there were some problems after merging master into this branch.

**Comment by coderaiser** [comment]
@nicolo-ribaudo looks like something wrong on `windows` with `node v14` in places that not related to this PR.

**Comment by nicolo-ribaudo** [comment]
It's a node bug, we are investigating it.

**Comment by coderaiser** [comment]
Any progress on this :)?

**Comment by JLHwung** [comment]
@coderaiser Can you rebase on `main`?

**Comment by coderaiser** [comment]
@JLHwung done :)

File: packages/babel-core/src/transformation/normalize-file.js [hunk 2]
```diff
@@ -83,6 +83,27 @@ export default function* normalizeFile(
...
+  if (ast) {
+    if (ast.type === "Program") {
+      ast = t.file(ast, [], []);
+    } else if (ast.type !== "File") {
+      throw new Error("AST root must be a Program or File node");
+    }
+
+    const { cloneInputAst } = options;
+    if (typeof cloneInputAst !== "boolean") {
+      throw Error("cloneInputAst should be a boolean");
```
Comment (JLHwung) [inline packages/babel-core/src/transformation/normalize-file.js:95]: Haven't `cloneInputAst` been validated in `packages/babel-core/src/config/validation/options.js`?

Comment (coderaiser) [reply packages/babel-core/src/transformation/normalize-file.js]: Removed additional check :).

**Comment by coderaiser** [comment]
Looks like `test262` takes too long :)

**Comment by nicolo-ribaudo** [comment]
Yeah sorry, we don't run test262 on PRs automatically since often it's not needed. We haven't merged this PR yet just because we are waiting for the next minor release.

**Pull request merged**